WIP INT-3: follow Link headers for pagination

rupa opened the pull request:
Applies to zones list, and get (for records)

WIP: to see if this approach is OK
TODO: tests, sorting?

Commit:
WIP INT-3: follow Link headers for pagination

Applies to zones list, and get (for records)

WIP: to see if this approach is OK
TODO: tests, sorting?

Commit:
WIP INT-3: FollowPagination setting and tests

Allow new pagination behavior to be turned off. This will allow users
to opt-out, if for some reason the new behavior doesn't align to
their needs. In the future, we may move to a finer-grained approach
to pagination, for now, it's on-and-automatic or off.

tests for pagination using a mocked API server

mburtless commented on rest/client.go:
```diff
@@ -74,10 +77,11 @@ func NewClient(httpClient Doer, options ...func(*Client)) *Client {
+		Endpoint:         endpoint,
+		RateLimitFunc:    defaultRateLimitFunc,
+		UserAgent:        defaultUserAgent,
+		FollowPagination: true,
```
Should we use a `defaultShouldFollow` var to contain the default value, instead of embedding it here?  Makes it easier to read if all defaults are kept in the same place instead of mixing embedding values and vars IMO.

rupa replied:
done

mburtless commented on rest/zone.go:
```diff
@@ -52,9 +64,55 @@ func (s *ZonesService) Get(zone string) (*dns.Zone, *http.Response, error) {
 	return &z, resp, nil
 }
 
+func (s *ZonesService) nextZones(zl *[]*dns.Zone, uri string) (*http.Response, error) {
```
Does `zl` need to be a pointer to a slice?  Slice values are effectively passed by reference in Go, since the value of a slice is just a pointer to an underlying array and two other ints [1].  Accordingly, passing a slice by value is very cheap and will simplify your code below.

[1] https://blog.golang.org/go-slices-usage-and-internals

rupa replied:
I couldn't get this to work, I think because we aren't returning `zl`, we're mutating it as a side effect.

mburtless replied:
Doh, I totally overlooked we weren't returning `zl`.  You're totally right.

mburtless commented on rest/zone.go:
```diff
@@ -52,9 +64,55 @@ func (s *ZonesService) Get(zone string) (*dns.Zone, *http.Response, error) {
+	}
+	for z := range tmpZl {
+		*zl = append(*zl, tmpZl[z])
+	}
```
`append` allows you to append a slice to another slice without manual iteration by using `...` to expand the second argument.

i.e. `zl = append(zl, tmpZl...)`

rupa replied:
done

mburtless commented on rest/zone.go:
```diff
@@ -52,9 +64,55 @@ func (s *ZonesService) Get(zone string) (*dns.Zone, *http.Response, error) {
+	// paginated response.
+	for r := range tmpZone.Records {
+		z.Records = append(z.Records, tmpZone.Records[r])
+	}
```
Same deal, this could be simplified to `z.Records = append(z.Records, tmpZone.Records...)`

rupa replied:
done

mburtless commented on rest/headers.go:
```diff
@@ -0,0 +1,89 @@
+
+		pieces := linkMatches[0]
+
+		link := &Link{URI: pieces[1], Extra: map[string]string{}}
```
IMO we should validate that we received a legit URI before returning it to protect against API returning bad values.

(i.e. https://play.golang.org/p/JngFarWPF2-, https://stackoverflow.com/questions/31480710/validate-url-with-standard-package-in-go).

mburtless commented on rest/zone_test.go:
```diff
@@ -0,0 +1,161 @@
+    "short_answers": ["4.4.4.4"],
+    "type":"A"
+  }]
+}`
```
Are package scoped vars necessary here.  Can these live inside `mockAPI`?  Alternatively, can `mockAPI` be generalized by taking it's configuration (i.e. `http.HandlerFunc`)  from it's caller so it can be reused in the future?

rupa replied:
not necessary but seemed more readable than having them inside `mockApi` and not very dangerous. Definitely don't mind moving them.

I thought about generalizing `mockAPI` more, and thought about ways of organizing the data it uses better, but prefer to wait until we have some other use cases. The repo could use some more tests and I'm happy to work on some :) for now, I kinda prefer it dumb-n-hardcoded like this, it's a simple place to start when we add more tests n stuff

mburtless reviewed (changes requested):
Implementation makes sense IMO, a couple of notes about types and places where code could be simplified or have stronger validation.

One thing I'm seeing is that `nextRecords` and `nextZones` are very similar.  Seems like the only difference is the type of the temp variable used to hold the results of the request and where they get appended.

I think you could factor this out into a more DRY implementation that lives in `client` by creating a func (i.e. `getNext`) that accepts a pointer to a temp var to hold results and a URI and performs the actual GET.  `nextRecords` and `nextZones` could just handle setting up the temp var and performing the append.  Could be useful if we have to reuse this behaviour for other endpoints in the future.

Commit:
WIP INT-3: refactor pagination

* headers.go: check that Link is a valid URI
* push more of the pagination logic to client
* use splat for loopless append
* waste a variable for default pagination behavior

Commit:
WIP INT-3: refactor `mockAPI`

yeah it's pretty dumb, but prefer to generalize/make it better when
actually writing more tests that want to use it.

mburtless commented on rest/client.go:
```diff
@@ -156,6 +167,26 @@ func (c Client) Do(req *http.Request, v interface{}) (*http.Response, error) {
 }
 
+// nextFunc knows how to get and parse additional info from uri into v.
+type nextFunc func(v *interface{}, uri string) (*http.Response, error)
```
Love it.  This solution feels very Go, like http.HandlerFunc.

kevin-cantwell replied:
Oops, resolved this by accident!

kevin-cantwell commented on rest/client.go:
```diff
@@ -74,10 +79,11 @@ func NewClient(httpClient Doer, options ...func(*Client)) *Client {
-		Endpoint:      endpoint,
-		RateLimitFunc: defaultRateLimitFunc,
-		UserAgent:     defaultUserAgent,
+		httpClient:       httpClient,
```
If you use `httpClient: http.DefaultClient` instead, you can drop the `httpClient Doer` argument and the default invocation of `NewClient` just becomes: `rest.NewClient()`

rupa replied:
This is definitely better, but a breaking change that doesn't seem worth it at the moment.

kevin-cantwell commented on rest/client.go:
```diff
@@ -156,6 +167,26 @@ func (c Client) Do(req *http.Request, v interface{}) (*http.Response, error) {
+type nextFunc func(v *interface{}, uri string) (*http.Response, error)
+
+// DoWithPagination Does, and follows Link headers for pagination.
+func (c Client) DoWithPagination(req *http.Request, v interface{}, f nextFunc) (*http.Response, error) {
```
I recommend you use a pointer receiver here `func (c *Client) ...`

As it's currently set up, I don't think it would cause any problems, but you may find it helps avoid problems down the road if any of your fields become non-pointer structs.

rupa replied:
done. huh, this guy copied off of `Do`, which is/was the odd one out. Seems like `Do` could/should also have a pointer receiver

kevin-cantwell commented on rest/client.go:
```diff
@@ -156,6 +167,26 @@ func (c Client) Do(req *http.Request, v interface{}) (*http.Response, error) {
+type nextFunc func(v *interface{}, uri string) (*http.Response, error)
+
+// DoWithPagination Does, and follows Link headers for pagination.
+func (c Client) DoWithPagination(req *http.Request, v interface{}, f nextFunc) (*http.Response, error) {
```
You have mixed exposed and unexposed types in this signature. I'd pick one and stick with it. However, is there any strong reason why this function should be exposed at all?

rupa replied:
good point. `Do` could probably also be unexposed. But it could be nice - from an SDK as toolkit perspective - to let users get at some of the guts if they want do do something fancy.

For consistency and least amount of change, inclined to keep `DoWithPagination` exposed and uppercase `NextFunc`

kevin-cantwell commented on rest/zone.go:
```diff
@@ -55,6 +65,32 @@ func (s *ZonesService) Get(zone string) (*dns.Zone, *http.Response, error) {
+	if err != nil {
+		return nil, err
+	}
+	zoneList := (*v).(*[]*dns.Zone)
```
Oof. I think this is convoluted and brittle. One has to think too hard about what's happening to `v` when invoking this as a `nextFunc` interface. 

Consider changing the signature of your pagination function to accept a page handler type:
`func (c *Client) DoWithPagination(req *http.Request, page func(*http.Response) error) error`

This allows the caller to handle each response explicitly, eg:
```go
if err := s.client.DoWithPagination(req, func(resp *http.Response) error {
  z, err := parseZone(resp)
  if err != nil {
    return err
  }
  zl = append(zl, z)
  return nil
}); err != nil {
  return nil, err
}
```

rupa replied:
yeah this is not the cleanest or most maintainable approach, but I think we wanna stick with it for now. Longer term I think we'd like to advocate/discuss some API changes that would allow for finer-grained pagination control for users, without exposing them to confusing implementation details.

kevin-cantwell commented on rest/zone.go:
```diff
@@ -21,7 +21,12 @@ func (s *ZonesService) List() ([]*dns.Zone, *http.Response, error) {
-	resp, err := s.client.Do(req, &zl)
+	var resp *http.Response
+	if s.client.FollowPagination == true {
+		resp, err = s.client.DoWithPagination(req, &zl, s.nextZones)
```
I think it's a problem if this function `List()` returns the raw http.Response, especially wrt pagination. Which page is it getting? The first? Last? Is it useful when paging is enabled?

rupa replied:
yeah good point. We probably don't need to return a Response, and shouldn't. Again tho, probably something we don't want to change right away due to breaking things.

kevin-cantwell replied:
It's debatable, but I think the pattern AWS uses with all their api interfaces is pretty flexible and maintainable: 
```
func (s *AWSService) ServiceCallXYZ(input * XYZInput) (*XYZOutput, error) {
  // all the http logic, retries, and pagination, etc.
}
```

rupa replied:
yeah, agreed. I don't think the httpResponse should get out of these funcs. I don't think anything uses it or needs it - but that change will have to wait for a breaking version. For now, `DoWithPagination` matches `Do` in that it returns the first response

kevin-cantwell commented on rest/zone.go:
```diff
@@ -55,6 +65,32 @@ func (s *ZonesService) Get(zone string) (*dns.Zone, *http.Response, error) {
+	if err != nil {
+		return nil, err
+	}
+	zone := (*v).(*dns.Zone)
```
Any type of casting like this is a code smell in Go. At the very least, you should be type checking here:
```
zone, ok := (*v).(*dns.Zone)
if !ok {
  return nil, errors.New(...)
}
```

rupa replied:
✔️

kevin-cantwell reviewed (commented):
Dropped a few comments, but one larger concern I'd like to bring up is the package name.

The name `rest` is not a very "Go" package name here. It's too broad. Are we exposing an implementation of the REST protocol in this package? If not, it should be `ns1rest` or `ns1api` or better yet, the public client should be exposed simply under the root package: `ns1.NewClient`.

See our RFC on this subject for more info: https://ns1inc.atlassian.net/browse/RFC-10

Commit:
WIP INT-3: changes from review

* Expose `NextFunc` type to users in case they want to write one.
* Change `Do` and `DoWithPagination` to have a pointer receiver,
  consistent with other methods on `Client`.
* Check/validate the interface casts in `nextRecords` and `nextZones`,
  and move the func definitions toward the bottom of the file.

Commit:
WIP INT-3: put back the non-pointer receiver for Do(WithPagination)

Prefer to keep Do the way it is, as it emphasizes that it can't
mutate Client, and we want DoWithPagination to match Do.
